pkg/changelog/changelog.go

```diff
@@ -0,0 +1,207 @@
+)
+
+// Regex patterns used by all Giant Swarm components
+const (
+	commonStartPattern = "(?m)^## \\[v?(?P<Version>\\d+\\.\\d+\\.\\d+)\\].*(?P<Date>\\d{4}-\\d{2}\\-\\d{2})$"
+	commonEndPattern   = "(?m)^\\[.*\\]:.*$"
+)
+
+// Parameters defining how to parse and extract release info about all known components
+var knownComponentParseParams = map[string]parseParams{
+	"app-operator": {
+		tag:       "https://github.com/giantswarm/app-operator/releases/tag/v{{.Version}}",
+		changelog: "https://raw.githubusercontent.com/giantswarm/app-operator/master/CHANGELOG.md",
+		start:     commonStartPattern,
+		end:       commonEndPattern,
+	},
+	"aws-operator": {
+		tag:       "https://github.com/giantswarm/aws-operator/releases/tag/v{{.Version}}",
+		changelog: "https://raw.githubusercontent.com/giantswarm/aws-operator/master/CHANGELOG.md",
+		start:     commonStartPattern,
+		end:       commonEndPattern,
+	},
+	"cert-operator": {
+		tag:       "https://github.com/giantswarm/cert-operator/releases/tag/v{{.Version}}",
+		changelog: "https://raw.githubusercontent.com/giantswarm/cert-operator/master/CHANGELOG.md",
+		start:     commonStartPattern,
+		end:       commonEndPattern,
+	},
+	"cluster-operator": {
+		tag:       "https://github.com/giantswarm/cluster-operator/releases/tag/v{{.Version}}",
+		changelog: "https://raw.githubusercontent.com/giantswarm/cluster-operator/master/CHANGELOG.md",
+		start:     commonStartPattern,
+		end:       commonEndPattern,
+	},
+	"kubernetes": {
+		tag:          "https://github.com/kubernetes/kubernetes/releases/tag/v{{.Version}}",
+		changelog:    "https://raw.githubusercontent.com/kubernetes/kubernetes/master/CHANGELOG/CHANGELOG-{{.Major}}.{{.Minor}}.md",
+		start:        "(?m)^# v?(?P<Version>\\d+\\.\\d+\\.\\d+)$",
+		intermediate: "(?m)^## Changes by Kind$",
+		end:          "(?m)^# .*$",
+	},
+	"etcd": {
+		tag:       "https://github.com/etcd-io/etcd/releases/tag/v{{.Version}}",
+		changelog: "https://raw.githubusercontent.com/etcd-io/etcd/master/CHANGELOG-{{.Major}}.{{.Minor}}.md",
+		start:     "(?m)^## \\[v?(?P<Version>\\d+\\.\\d+\\.\\d+)\\].*$",
+		end:       "(?m)^## .*$",
+	},
+	"aws-cni": {
+		tag:       "https://github.com/aws/amazon-vpc-cni-k8s/releases/tag/v{{.Version}}",
+		changelog: "https://raw.githubusercontent.com/aws/amazon-vpc-cni-k8s/master/CHANGELOG.md",
+		start:     "(?m)^## v?(?P<Version>\\d+\\.\\d+\\.\\d+)$",
+		end:       "(?m)^##? .*$",
+	},
+}
+
+type parseParams struct {
+	tag          string
+	changelog    string
+	start        string
+	intermediate string
+	end          string
+}
+
+// Data about a component passed into templates that depend on versions
+type versionTemplateData struct {
+	Major   uint64
+	Minor   uint64
+	Version string
+}
+
+// Data about a particular component version returned from parsing a changelog
+type Version struct {
+	Link    string
+	Name    string
+	Content string
+}
+
+func ParseChangelog(componentName, componentVersion string) (*Version, error) {
```

Comment:
This is one of the core functions. Returns the changes for a specific version of an upstream component by parsing its changelogs.

Reply:
Dios mio - this is certainly a function. Fine for now but maybe we can abstract some parts of it.

Reply:
Indeed. I cleaned it up a bit. There's plenty of room for improvement (and unit tests), but it does work for the most part.